NetCasbin/Rbac/DefaultRoleManager.cs

```diff
@@ -1,163 +1,271 @@
+        private void AddLink(string name1, string name2, string domain = null)
+        {
+            domain ??= _defaultDomain;
+            var roles = _allDomains.GetOrAdd(domain,
+                new ConcurrentDictionary<string, Role>());
+
+            Role role1 = roles.GetOrAdd(name1, new Role(name1));
+            Role role2 = roles.GetOrAdd(name2, new Role(name2));
+            role1.AddRole(role2);
+        }
+
+        private void DeleteLink(string name1, string name2, string domain = null)
         {
-            if (domain.Length == 1)
+            domain ??= _defaultDomain;
+            var roles = _allDomains.GetOrAdd(domain,
+                new ConcurrentDictionary<string, Role>());
+
+            if (roles.ContainsKey(name1) is false
+                || roles.ContainsKey(name2) is false)
             {
-                name = domain[0] + "::" + name;
+                ThrowHelper.ThrowOneOfNamesNotFoundException();
             }
-            else if (domain.Length > 1)
+
+            Role role1 = roles.GetOrAdd(name1, new Role(name1));
+            Role role2 = roles.GetOrAdd(name2, new Role(name2));
+            role1.DeleteRole(role2);
+        }
+
+        public virtual void Clear()
+        {
+            _allDomains.Clear();
+        }
+
+        private static bool HasRole(ConcurrentDictionary<string, Role> roles, string name,
+            Func<string, string, bool> matchingFunc = null)
+        {
+            if (matchingFunc is null)
             {
-                throw new Exception("error: domain should be 1 parameter");
+                return roles.ContainsKey(name);
             }
-            if (!HasRole(name))
+
+            ICollection<string> keys = roles.Keys;
+            return keys.Any(key => matchingFunc(name, key));
+        }
+
+        private static Role CreateRole(ConcurrentDictionary<string, Role> roles, string name,
+            Func<string, string, bool> matchingFunc = null)
+        {
+            Role role = roles.GetOrAdd(name, new Role(name));
+
+            if (matchingFunc is null)
             {
-                //return [];
+                return role;
             }
 
-            var users = _allRoles.Values.Where(x => x.HasDirectRole(name)).Select(x => x.Name);
-            if (domain.Length == 1)
+            ICollection<string> keys = roles.Keys;
+            foreach (string key in keys)
             {
-                users = users.Select(n => n.Substring(domain[0].Length + 2, n.Length));
+                if (name == key || matchingFunc(name, key) is false)
+                {
+                    continue;
+                }
+
+                Role matchingRole = roles.GetOrAdd(key, new Role(key));
+                role.AddRole(matchingRole);
             }
-            return users.ToList();
+            return role;
         }
 
-
-        private Role CreateRole(string name)
+        public ConcurrentDictionary<string, Role> GenerateTempRoles(string domain)
```

Comment:
This method is not test coveraged, you can add a simple test case so I can debug and make sure the logic is right, only reading code is too hard for me.

Reply:
You are right, We need to add more tests.

Reply:
Tests have been included in the last commit.